c10/core/impl/cow/state_machine.cpp

```diff
@@ -0,0 +1,240 @@
+    // shadow storage on a tensor if its own storage created it. We
+    // don't check for the specific matching of the storage and shadow
+    // storage, but we do check that the presence relationship holds.
+    //
+    // TODO: This check should probably be here, but it is actually
+    // triggering in the StaticRuntime.autogen_inner test.
+    //
+    // TORCH_INTERNAL_ASSERT(maybe_shadow_storage == nullptr);
+    return;
+  }
+
+  impl->maybe_bump(maybe_shadow_storage);
+}
+
+auto cow::StateMachine::simulate_lazy_copy(
+    cow::ShadowStorage* maybe_shadow_storage)
+    -> intrusive_ptr<cow::ShadowStorage> {
+  return ensure_initialized().simulate_lazy_copy(maybe_shadow_storage);
+}
+
+auto cow::StateMachine::maybe_get_impl() -> cow::StateMachine::Impl* {
+  return get_impl([] { return nullptr; });
+}
+
+auto cow::StateMachine::ensure_initialized() -> cow::StateMachine::Impl& {
+  return *get_impl([this]() -> c10::optional<cow::StateMachine::Impl*> {
+    Rep initial_rep(StateId::initial);
+    Rep initializing_rep(cow::StateMachine::StateId::initializing);
+    // If we fail to exchange, it means that someone else is
+    // initializing.
+    if (!state_rep_.compare_exchange_strong(initial_rep, initializing_rep)) {
+      return c10::nullopt;
+    }
+
+    auto impl = std::make_unique<cow::StateMachine::Impl>();
+    Rep new_rep(*impl);
+    TORCH_INTERNAL_ASSERT(
+        new_rep.as_state_id() == cow::StateMachine::StateId::active);
+    // We won the race to initializing: we will be the only thread
+    // that may transition to active.
+    TORCH_INTERNAL_ASSERT(
+        state_rep_.compare_exchange_strong(initializing_rep, new_rep));
+    return c10::make_optional(impl.release()); // owned by state_rep_ now
+  });
+}
+
+template <typename OnInitialFunc>
+auto cow::StateMachine::get_impl(const OnInitialFunc& on_initial_func)
+    -> StateMachine::Impl* {
+  bool seen_initializing = false;
+  do {
+    Rep state_rep = state_rep_;
+    switch (state_rep.as_state_id()) {
+      case StateId::initial: {
+        TORCH_INTERNAL_ASSERT(!seen_initializing);
+        c10::optional<Impl*> impl = on_initial_func();
+        if (!impl.has_value()) {
+          continue;
+        }
+        return *impl;
+      }
+
+      case StateId::initializing:
+        seen_initializing = true;
+        // Wait for initialization.
+        continue;
+
+      case StateId::active:
+        return &state_rep.as_state_machine_impl();
+    }
+  } while (true);
+}
+
+cow::StateMachine::Rep::Rep(cow::StateMachine::StateId state_id)
+    : rep_(static_cast<std::uintptr_t>(state_id)) {
+  TORCH_INTERNAL_ASSERT(as_state_id() == state_id);
+}
+
+namespace {
+
+// Implementations of std::bit_cast() from C++ 20.
+//
+// This is a less sketchy version of reinterpret_cast.
+//
+// See https://en.cppreference.com/w/cpp/numeric/bit_cast for more
+// information as well as the source of our implementations.
```

Comment:
Put it in c10/utils?

Reply:
this 3 diff stack addresses that: D44712290